numpy/_core/src/multiarray/lowlevel_strided_loops.c.src

```diff
@@ -869,6 +894,65 @@ NPY_NO_EXPORT PyArrayMethod_StridedLoop *
     #define GCC_CAST_OPT_LEVEL NPY_GCC_OPT_3
 #endif
 
+#define _RETURN_SAME_VALUE_FAILURE \
+        npy_gil_error(PyExc_ValueError, "could not cast 'same_value' @name1@ to @name2@"); \
+        return -1
+
+#if !@is_bool2@
+/*
+ * Check various modes of failure to accurately cast src_value to dst
+ */
+static GCC_CAST_OPT_LEVEL int
+@prefix@_check_same_value_@name1@_to_@name2@(@rtype1@ src_value) {
+
+    int src_isnan = isnan((long double)_TO_RTYPE1(src_value));
+    int src_isinf = isinf((long double)_TO_RTYPE1(src_value));
+    /*
+     * 1. Check that the src does not overflow the dst.
+     *    Ignore inf since it does not compare well
+     *    This is complicated by a warning that, for instance, int8 cannot
+     *    overflow int64max
+     */
+#   ifdef __GNUC__
+#   pragma GCC diagnostic push
+#   ifdef __clang__
+#   pragma GCC diagnostic ignored "-Wtautological-constant-out-of-range-compare"
+#   endif
+#   pragma GCC diagnostic ignored "-Wtautological-compare"
+#   endif
+#if (@is_float2@ || @is_emu_half2@ || @is_double2@ || @is_native_half2@)
+    if (!src_isinf) {
+#endif
+#      if !@is_bool1@
+       if (!src_isnan && _TO_RTYPE1(src_value) > @type2max@) {
+            _RETURN_SAME_VALUE_FAILURE;
+       }
+#      if !@is_unsigned1@
+       if (!src_isnan && _TO_RTYPE1(src_value) < @type2min@) {
+            _RETURN_SAME_VALUE_FAILURE;
+       }
+#      endif
+#      endif /* !is_bool1 */
+#   if (@is_float2@ || @is_emu_half2@ || @is_double2@ || @is_native_half2@)
+    }
+#   endif
+    /* 2. Check inf, nan with a non-float dst */
+#   if (!@is_float2@ && !@is_emu_half2@ && !@is_double2@ && !@is_native_half2@)
+        if (src_isnan || src_isinf) {
+        _RETURN_SAME_VALUE_FAILURE;
+    }
+#   endif
+    /* 3. Check that the value can round trip exactly */
+    if (!src_isnan && !src_isinf && src_value != _ROUND_TRIP(src_value)) {
```

Comment:
Could skip for all integers (the range check is what's important).  But I guess there is a decent chance the compiler will do it right.